File: Behavioral/Lab20/Typescript/src/Entities/room.ts
import { RealEstate } from './real_estate';
import { Customer } from './customer';
import { Realtor } from './realtor';
export class Room extends RealEstate {
    public heatingType: string;
    public isGoodNeighbors: boolean;
    constructor();
    constructor(address: string, description: string, area: number, price: number, status: string);
    constructor(address?: string, description?: string, area?: number, price?: number, status?: string) {
        if (address && description && area && price && status) {
            super(address, description, area, price, status);
        } else {
            super("Dungeon", "tra ta ta", 50, 1000, "New");
        }
    }

    public toString(): string {
        return "Room:\n" + super.toString();
    }

    public clone(): Room {
        const clone = Object.assign(Object.create(Object.getPrototypeOf(this)), this);
        clone.realtor = this.realtor ? Realtor.newRealtorFromObject(this.realtor) : null;
        return clone;
    }

    public shallowCopy(): Room {
        return Object.assign(Object.create(Object.getPrototypeOf(this)), this);
    }

    public reset(): void {
        super.reset();
        this.heatingType = "";
    }
}

File: Behavioral/Lab20/C#/Entities/Room.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.AccessControl;
using System.Text;
using System.Threading.Tasks;

namespace C_.Entities
{
    public class Room : RealEstate
    {
        public string HeatingType { get; set; }
        public bool IsGoodNeighbors { get; set; } = false;
        public Room():base("Dungeon", "tra ta ta", 50, 1000, "New"){
        }
        public Room(string address, string description, double area, double price, string status) :
            base(address, description, area, price, status)
        {   
        }

        public override string ToString(){
            return "Room:\n"+ base.ToString();
        }

        public override Room Clone()
        {
            var clone = (Room)MemberwiseClone();
            clone.Realtor = Realtor == null ? Realtor : new Realtor(Realtor);
            return clone;
        }

        public override Room ShallowCopy()
        {
            return (Room)MemberwiseClone();
        }

        public override void Reset()
        {
            base.Reset();
            HeatingType = "";
        }
    }
}
